kernel/alarm.c

```diff
@@ -0,0 +1,251 @@
+
+	while (cp) {
+		struct k_alarm *cap = alarm_due_by(cp, now);
+
+		if (!cap) {
+			break;
+		}
+		cap->state = K_ALARM_READY;
+		++rv;
+		cp = sys_dlist_peek_next_no_check(&scheduled_list, cp);
+	}
+	if ((rv == 0)
+	    && !sys_dlist_is_empty(&ready_list)) {
+		rv = -1;
+	}
+
+	if (cp) {
+		sys_dlist_split(&ready_list, &scheduled_list, cp);
+	} else {
+		sys_dlist_join(&ready_list, &scheduled_list);
+	}
+
+	k_spin_unlock(&lock, key);
+	return rv;
+}
+
+void k_alarm_process_ready_(void)
+{
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	do {
+		sys_dnode_t *cp = sys_dlist_get(&ready_list);
+
+		if (!cp) {
+			break;
+		}
+
+		struct k_alarm *cap = CONTAINER_OF(cp, struct k_alarm, node);
+
+		cap->state = K_ALARM_ACTIVE;
+		if (cap->handler_fn) {
+			k_spin_unlock(&lock, key);
+			cap->handler_fn(cap, cap->user_data);
+			key = k_spin_lock(&lock);
+		}
+		if (cap->state == K_ALARM_ACTIVE) {
+			cap->state = K_ALARM_UNSCHEDULED;
+		}
+	} while (true);
+	k_spin_unlock(&lock, key);
+}
+
+int z_impl_k_alarm_schedule(struct k_alarm *alarm,
+			    u32_t deadline,
+			    u32_t flags)
+{
+	int rv = -EINVAL;
+	bool update_deadline = false;
+
+	if (!alarm) {
+		goto out;
+	}
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	bool may_resched = true
+		&& (alarm->state == K_ALARM_SCHEDULED)
+		&& (flags & (K_ALARM_FLAG_REPLACE | K_ALARM_FLAG_IF_SOONER));
+
+	if (sys_dnode_is_linked(&alarm->node)
+	    && !may_resched) {
+		rv = -EINVAL;
+		goto unlock;
+	}
+	if ((alarm->state != K_ALARM_UNSCHEDULED)
+	    && (alarm->state != K_ALARM_ACTIVE)
+	    && (alarm->state != K_ALARM_CANCELLED)
+	    && !may_resched) {
+		rv = -EBUSY;
+		goto unlock;
+	}
+
+	u32_t now = k_cycle_get_32();
+	s32_t delay = deadline - now;
+
+	if (may_resched) {
+		if (K_ALARM_FLAG_REPLACE & flags) {
+			/* Unconditional reschedule */
+		} else if ((K_ALARM_FLAG_IF_SOONER & flags)
+			   && (delay >= (s32_t)(alarm->deadline - now))) {
+			/* Not sooner; keep unchanged */
+			rv = 1;
+			goto unlock;
+		}
+		sys_dlist_remove(&alarm->node);
+	}
+
+	alarm->deadline = deadline;
+	if (delay > 0) {
+		alarm->state = K_ALARM_SCHEDULED;
+		update_deadline = link_alarm(&scheduled_list, alarm, now);
+		rv = 1;
+	} else if (K_ALARM_FLAG_ERROR_IF_LATE & flags) {
+		rv = -EINVAL;
+	} else {
+		alarm->state = K_ALARM_READY;
+		update_deadline = link_alarm(&ready_list, alarm, now);
+		rv = 0;
+	}
+
+unlock:
+	k_spin_unlock(&lock, key);
+out:
+	if (update_deadline) {
+		z_alarm_update_deadline();
+	}
+
+	return rv;
+}
+
+int z_impl_k_alarm_cancel(struct k_alarm *alarm)
+{
+	int rv = -EINVAL;
+	bool update_deadline = false;
+
+	if (!alarm) {
+		goto out;
+	}
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	if (alarm->state == K_ALARM_SCHEDULED) {
+		update_deadline = (&alarm->node
+				   == sys_dlist_peek_head(&scheduled_list));
+		rv = 1;
+		sys_dlist_remove(&alarm->node);
+	} else if (alarm->state == K_ALARM_READY) {
+		sys_dlist_remove(&alarm->node);
```

Comment:
Doesn't have to check.  If it's `READY` it's on the ready queue and hasn't been processed, so it can simply be removed.

Reply:
the case where we see that alarm is `READY` happens when we interrupted RTC interrupt so we are in higher priority interrupt. We may as well interrupted RTC interrupt before ready state is determined and alarm also may be already late and that case is not covered and from user perspective they are the same.

Reply:
The alarm is also observably `READY` when multiple alarms reach their deadline in the same timer ISR, and one callback tries to cancel another alarm while it's still on the queue.

I think this can be resolved by clarifying that "reached its deadline" does not just mean that the RTC counter has advanced to the deadline (which we can't easily observe), but that the alarm and timer infrastructure has recognized that event has happened (which we can observe, because it's reflected in the alarm state being `READY`).

I will reword the documentation to make that clear.

This relates to your comment about the return value from `k_alarm_cancel()`.  As a user what I want to know is what the *state* of the alarm was (distinguish `SCHEDULED` from `READY`), not a difference in ticks that was captured at some arbitrary point in the past.  Any such difference is irrelevant since we don't know how long ago it was calculated.